### PR TITLE
Support SSL with client auth for Mongo

### PR DESCRIPTION
**Pull request by metamben**
Part of epic #22889.

- [x] Generate keys and certificates for the CA, Mongo (the server) and Metabase (the client)
- [x] Create/configure a Mongo Docker image requiring client-side TLS certificate
- [x] Implement getting the client-side TLS configuration from the Metabase UI
- [x] Implement getting the client-side TLS configuration from Metabase secret store
- [x] Add backend tests for client-side SSL and get them pass locally
- [x] Check in keys and certificates for the Mongo server so that CircleCI can access them
- [x] Check in keys and certificates for Metabase so that they are available during testing
- [x] Adjust `.circleci/config.yaml` to use the new image or new config options
- [x] Configure the frontend to allow client-side SSL configuration for Mongo
- [x] Replace `textFile` property for client certificate with `text` to make it symmetrical to the field of the server certificate
- [x] ~Add end-to-end test~

### CODE REVIEW

File: src/metabase/driver/util.clj [hunk 1]
```diff
@@ -356,6 +357,9 @@
                 (assoc :visible-if v-ifs*))))
          final-props)))
 
+(def ^:private ^Base64$Decoder base64-decoder
+  (Base64/getDecoder))
```
Comment (noahmoss) [inline src/metabase/driver/util.clj:361]: Similar comment here as above, about using the `metabase.util` decoder if possible.

Comment (noahmoss) [inline src/metabase/driver/util.clj:361]: Similar comment here as above, about using the `metabase.util` decoder if possible.

File: src/metabase/driver/util.clj [hunk 2]
```diff
@@ -459,12 +463,47 @@
...
+(defn- parse-rsa-key
+  "Parses an RSA private key from the PEM string `key-string`."
+  ^PrivateKey [key-string]
+  (let [algorithm (or (key-type key-string) "RSA")
+        key-base64 (-> key-string
+                       (str/replace #"^-----BEGIN (?:(\p{Alnum}+) )?PRIVATE KEY-----\n" "")
+                       (str/replace #"\n-----END (?:(\p{Alnum}+) )?PRIVATE KEY-----\s*$" "")
+                       (str/replace #"\s" ""))
+        decoded (.decode base64-decoder key-base64)
+        key-factory (KeyFactory/getInstance algorithm)] ; TODO support other algorithms
```
Comment (noahmoss) [inline src/metabase/driver/util.clj:479]: Won't this blow up if an unknown algorithm is parsed from the PEM file?

**Review by noahmoss — COMMENTED** [review]
I don't have a _ton_ of domain knowledge around the SSL & secrets stuff but I left a bunch of random comments on code things I noticed in an initial pass. Happy to take a look again once it's published.

File: modules/drivers/mongo/src/metabase/driver/mongo/util.clj [hunk 3]
```diff
@@ -93,25 +105,50 @@
   [user pass host dbname authdb]
   (format "mongodb+srv://%s:%s@%s/%s?authSource=%s" user pass host dbname authdb))
 
+(def ^:private ^Base64$Decoder base64-decoder
+  (Base64/getDecoder))
```
[the remark this answers is not in the document]
Comment (metamben) [reply modules/drivers/mongo/src/metabase/driver/mongo/util.clj:109]: I think it's a good idea to have just one place for this. I tend not to use extra dependencies if I can avoid it, so I'll make use of the built-in decoder.

File: modules/drivers/mongo/src/metabase/driver/mongo/util.clj [hunk 4]
```diff
@@ -93,25 +105,50 @@
...
+      (String. StandardCharsets/UTF_8)))
+
+(defn- get-secret [details secret-property]
+  (let [value-key (keyword (str secret-property "-value"))
+        options-key (keyword (str secret-property "-options"))
+        path-key (keyword (str secret-property "-path"))
+        id-key (keyword (str secret-property "-id"))
+        id (id-key details)
+        value (if id
+                (String. ^bytes (:value (secret/Secret id)) StandardCharsets/UTF_8)
```
[the remark this answers is not in the document]
Comment (metamben) [reply modules/drivers/mongo/src/metabase/driver/mongo/util.clj:123]: This was a conscious choice. The constant is in use elsewhere too, if I remember correctly.

File: src/metabase/driver/util.clj (hunk 1 above)
Comment (metamben) [reply src/metabase/driver/util.clj:361]: `(Base64/getDecoder)` was in use before here, I've just extracted it. But anyway, I'll have a look if I can move it over.

File: modules/drivers/mongo/src/metabase/driver/mongo/util.clj [hunk 5]
```diff
@@ -93,25 +105,50 @@
...
 
+(def ^:private ^Base64$Decoder base64-decoder
+  (Base64/getDecoder))
+
+(defn- decode-uploaded [uploaded-data]
+  (-> base64-decoder
+      (.decode (str/replace uploaded-data #"^data:[^;]+;base64," ""))
+      (String. StandardCharsets/UTF_8)))
+
+(defn- get-secret [details secret-property]
```
[the remark this answers is not in the document]
Comment (metamben) [reply modules/drivers/mongo/src/metabase/driver/mongo/util.clj:116]: I move it to `metabase.models.secret`.

**Comment by mazameli** [comment]
@noahmoss, looks like this is ready for review when you get some time. 🙏🏼

**Review by noahmoss — APPROVED** [review]
Seems reasonable, with the caveat again that I don't have a super deep knowledge of this area and so very well may not be able to spot all issues.

**Review by escherize — APPROVED** [review]
I think it looks great. Thanks for this @metamben. I only had a question about the metaca.crt expiring in about 2 weeks.

File: test_resources/ssl/mongo/metaca.crt [hunk 6]
```diff
@@ -0,0 +1,22 @@
+-----BEGIN CERTIFICATE-----
```
[the remark this answers is not in the document]
Comment (metamben) [reply test_resources/ssl/mongo/metaca.crt:1]: Good catch! The intention was to create all certificates for ~10 years.

**Review by camsaul — APPROVED** [review]

**Pull request merged**